src/rust/src/x509.rs

```diff
@@ -391,6 +411,296 @@ fn load_der_x509_certificate(py: pyo3::Python<'_>, data: &[u8]) -> PyAsn1Result<
+            "ObjectIdentifier",
+            (self.raw.borrow_value().signature_alg.oid.to_string(),),
+        )
+    }
+
+    fn public_bytes<'p>(
+        &self,
+        py: pyo3::Python<'p>,
+        encoding: &pyo3::PyAny,
+    ) -> pyo3::PyResult<&'p pyo3::types::PyBytes> {
+        let encoding_class = py
+            .import("cryptography.hazmat.primitives.serialization")?
+            .getattr("Encoding")?;
+
+        let result = asn1::write_single(self.raw.borrow_value());
+        if encoding == encoding_class.getattr("DER")? {
+            Ok(pyo3::types::PyBytes::new(py, &result))
+        } else if encoding == encoding_class.getattr("PEM")? {
+            let pem = pem::encode_config(
+                &pem::Pem {
+                    tag: "CERTIFICATE REQUEST".to_string(),
+                    contents: result,
+                },
+                pem::EncodeConfig {
+                    line_ending: pem::LineEnding::LF,
+                },
+            )
+            .into_bytes();
+            Ok(pyo3::types::PyBytes::new(py, &pem))
+        } else {
+            Err(pyo3::exceptions::PyTypeError::new_err(
+                "encoding must be Encoding.DER or Encoding.PEM",
+            ))
+        }
+    }
+
+    fn get_attribute_for_oid<'p>(
+        &self,
+        py: pyo3::Python<'p>,
+        oid: &pyo3::PyAny,
+    ) -> pyo3::PyResult<&'p pyo3::PyAny> {
+        // TODO: Convert ObjectIdentifier in python to a rust class so this can be more efficient
+        let oid_str = oid
+            .getattr("dotted_string")?
+            .downcast::<pyo3::types::PyString>()?
+            .to_string();
+        let rust_oid = asn1::ObjectIdentifier::from_string(&oid_str).unwrap();
+        for attribute in self.raw.borrow_value().csr_info.attributes.clone() {
+            if rust_oid == attribute.type_id {
+                check_attribute_length(attribute.values.clone())?;
+                if let Some(val) = attribute.values.clone().next() {
+                    // We allow utf8string, printablestring, and ia5string at this time
+                    if val.tag() == asn1::Utf8String::TAG
+                        || val.tag() == asn1::PrintableString::TAG
+                        || val.tag() == asn1::IA5String::TAG
+                    {
+                        return Ok(pyo3::types::PyBytes::new(py, val.data()));
+                    } else {
+                        return Err(pyo3::exceptions::PyValueError::new_err(format!(
+                            "OID {} has a disallowed ASN.1 type: {}",
+                            oid,
+                            val.tag()
+                        )));
+                    }
+                }
+            }
+        }
+        Err(pyo3::PyErr::from_instance(
+            py.import("cryptography.x509")?.call_method1(
+                "AttributeNotFound",
+                (
+                    format!("No {} attribute was found", rust_oid.to_string()),
```

Comment:
If the extract thing works, you can ust use that value instead of converting rust_oid back to a string.